src/hotspot/share/gc/shared/classUnloadingContext.cpp

```diff
@@ -0,0 +1,177 @@
+  _cld_head(nullptr),
+  _num_nmethod_unlink_workers(num_workers),
+  _unlinked_nmethods(nullptr),
+  _lock_codeblob_free_separately(lock_codeblob_free_separately) {
+
+  assert(num_workers > 0, "must be");
+
+  _unlinked_nmethods = NEW_C_HEAP_ARRAY(NMethodSet*, num_workers, mtGC);
+  for (uint i = 0; i < num_workers; ++i) {
+    _unlinked_nmethods[i] = new NMethodSet();
+  }
+}
+
+DefaultClassUnloadingContext::~DefaultClassUnloadingContext() {
+  for (uint i = 0; i < _num_nmethod_unlink_workers; ++i) {
+    delete _unlinked_nmethods[i];
+  }
+  FREE_C_HEAP_ARRAY(NMethodSet*, _unlinked_nmethods);
+}
+
+bool DefaultClassUnloadingContext::has_unloaded_classes() const {
+  return _cld_head != nullptr;
+}
+
+void DefaultClassUnloadingContext::register_unloading_class_loader_data(ClassLoaderData* cld) {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+
+  cld->unload();
+
+  cld->set_unloading_next(_cld_head);
+  _cld_head = cld;
+}
+
+void DefaultClassUnloadingContext::purge_class_loader_data() {
+  for (ClassLoaderData* cld = _cld_head; cld != nullptr;) {
+    assert(cld->is_unloading(), "invariant");
+
+    ClassLoaderData* next = cld->unloading_next();
+    delete cld;
+    cld = next;
+  }
+}
+
+void DefaultClassUnloadingContext::classes_unloading_do(void f(Klass* const)) {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+  for (ClassLoaderData* cld = _cld_head; cld != nullptr; cld = cld->unloading_next()) {
+    assert(cld->is_unloading(), "invariant");
+    cld->classes_do(f);
+  }
+}
+
+void DefaultClassUnloadingContext::register_unlinked_nmethod(nmethod* nm) {
+  assert(_context != nullptr, "no context set");
+
+  assert(!nm->is_unlinked(), "Only register for unloading once");
+  assert(_num_nmethod_unlink_workers == 1 || Thread::current()->is_Worker_thread(), "must be worker thread if parallel");
+
+  uint worker_id = _num_nmethod_unlink_workers == 1 ? 0 : WorkerThread::worker_id();
+  assert(worker_id < _num_nmethod_unlink_workers, "larger than expected worker id %u", worker_id);
+
+  _unlinked_nmethods[worker_id]->append(nm);
+
+  nm->set_is_unlinked();
+}
+
+void DefaultClassUnloadingContext::purge_nmethods() {
+  assert(_context != nullptr, "no context set");
+
+  size_t freed_memory = 0;
+
+  for (uint i = 0; i < _num_nmethod_unlink_workers; ++i) {
+    NMethodSet* set = _unlinked_nmethods[i];
+    for (int j = 0; j < set->length(); ++j) {
+      nmethod* nm = set->at(j);
+      freed_memory += nm->size();
+      nm->purge(false /* free_code_cache_data */);
+    }
+  }
+
+  CodeCache::maybe_restart_compiler(freed_memory);
+}
+
+void DefaultClassUnloadingContext::free_code_blobs() {
+  assert(_context != nullptr, "no context set");
+
+  // Sort nmethods before freeing to benefit from optimizations. If Nmethods were
+  // collected in parallel, use a new temporary buffer for the result, otherwise
+  // sort in-place.
+  NMethodSet* nmethod_set = nullptr;
+
+  bool is_parallel = _num_nmethod_unlink_workers != 1;
```

Comment:
```suggestion
  bool is_parallel = _num_nmethod_unlink_workers > 1;
```
Seems more intuitive.